Introduce a send_delay for pilight component

janLo opened the pull request:
**Description:**
This fixes an issue with pilight used with the [pilight USB Nano](https://github.com/pilight/pilight-usb-nano). If such a device is used, you need delays between sending codes to the transmitter. Otherwise it will skip random codes to transmit. This condition of mass-transmit is hit for example if a group of switches is turned off or on. Without a delay, random devices will fail to switch.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1325

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
# Example configuration.yaml entry
pilight: 
  host: 127.0.0.1
  port: 5000
  send_delay: 0.4
  whitelist:  # optional
    protocol:
      - daycom
      - intertechno
    id:
      - 42
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [ ] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

If pilight is used with a "pilight USB Nano" between the daemon and the
hardware, we must use a delay between sending multiple signals.
Otherwise the hardware will just skip random codes. We hit this
condition for example, if we switch a group of pilight switches on or
off. Without the delay, random switch signals will not be transmitted by
the RF transmitter.

As this seems not necessary, if the transmitter is directly connected
via GPIO, we introduce a optional configuration to set the delay.

janLo committed:
Introduce a send_delay for pilight component

If pilight is used with a "pilight USB Nano" between the daemon and the
hardware, we must use a delay between sending multiple signals.
Otherwise the hardware will just skip random codes. We hit this
condition for example, if we switch a group of pilight switches on or
off. Without the delay, random switch signals will not be transmitted by
the RF transmitter.

As this seems not necessary, if the transmitter is directly connected
via GPIO, we introduce a optional configuration to set the delay.

janLo commented:
The tests for python 3.4 are failing in an completely unrelated component. I cannot see how my changes have caused this.

DavidLP commented:
Looks fine to me. It would be nice to have a unit test checking if a config with `send_delay != 0.0` works. [This new line](https://coveralls.io/builds/8511187/source?filename=homeassistant%2Fcomponents%2Fpilight.py#L90) is the only one not called during tests.

janLo committed:
Add unittests for pilight send_delay handling

This adds an unittest to test the delayed calls to the send_code
service.

janLo commented:
Added two unittests to verify the call to time.sleep() if a delay is specified (and no call if its zero or less).

janLo commented on homeassistant/components/pilight.py:
```diff
@@ -79,6 +86,8 @@ def send_code(call):
         try:
             pilight_client.send_code(message_data)
+            if send_delay > 0.0:
+                time.sleep(send_delay)
```
So I have to track when the last call took place and manage a queue for my own to calculate the timepoint for the call correctly?

janLo replied:
@balloob Might this be a working solution: https://gist.github.com/janLo/5dba266f87ed3cdc0699720212d73757 ?

I'm not used to the async programming in python.

janLo replied:
The question is, how can I know how many calls are in the queue and since when to calculate the delay correctly? So my solution is, to manage a queue of actions within the service call. The first call runs directly, any subsequent service call will be put in the action queue. The actions itself schedule then the next action into the HA mainloop at the correct time until all queued actions are completed.

janLo committed:
Add a method to throttle calls to services

This adds CallRateDelayThrottle. This is a class that provides an
decorator to throttle calls to services. Instead of the Throttle in
homeassistant.util it does this by delaying all subsequent calls
instead of just dropping them. Dropping of calls would be bad if we
call services to actual change the state of a connected hardware (like
rf controlled power plugs).

Ihe delay is done by rescheduling the call using
track_point_in_utc_time from homeassistant.helpers.event so it should
not block the mainloop at all.

janLo committed:
Add unittests for CallRateDelayThrottle

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo commented:
I refactored the changes to use the homeassistant.helpers.event.track_point_in_utc_time like @balloob suggested. I implemented the delay handling in the service helper module. All changes are unittested.

janLo commented:
And a last update to make the linters happy.

balloob commented on homeassistant/helpers/service.py:
```diff
@@ -113,3 +117,58 @@ def extract_entity_ids(hass, service_call):
+
+
+# pylint: disable=too-few-public-methods
+class CallRateDelayThrottle(object):
```
I think that this might be a nice addition to the HA core in the future but in that case it should be async. Let's just keep this for now embedded in the pilight component.

janLo replied:
@balloob, I would prefer to just make it async.  Just tell me how I can achieve that. I have to change it anyway to get it merged.

balloob replied:
It's not very useful to make a platform use async if it uses a library that is not using async.

DavidLP replied:
It is on my TODO list to make the pilight library async.

janLo replied:
OK.  I'll put throttling  the code directly in the component as soon I've time and try to play around with async in a seperate branch.  I'll tell you if I've done any significant progress.  Depends all a bit on how much time I have.

janLo commented on homeassistant/helpers/service.py:
```diff
@@ -113,3 +117,58 @@ def extract_entity_ids(hass, service_call):
+    an overall throttling.
+
+    As this uses track_point_in_utc_time to schedule delayed executions
+    it should be async friendly.
```
OK.  What do I have to do to make it async friendly?  Just change it to the async_track_point_in_time version and remove the lock? As long as it is guaranteed that services are only called through the main loop the lock should not be necessary.

janLo commented:
Moved the delay code (including its unittests) into the pilight component (for now).

balloob commented:
Thanks for being flexible 👍 

Looks good 🐬
